src/semantics.rs

```diff
@@ -159,6 +159,142 @@ impl Validate for Expression {
     }
 }
 
+impl FieldCounter for Route {
+    fn add_to_counter(&self, map: &mut HashMap<String, usize>) {
+        for item in &self.stack {
+            match item {
+                RouteTerm::LogicalOperator(_op) => {
+                    // need to do nothing here
+                }
+                RouteTerm::Predicate(p) => {
+                    *map.entry(p.lhs.var_name.clone()).or_default() += 1;
+                }
+            }
+        }
+    }
+
+    fn remove_from_counter(&self, map: &mut HashMap<String, usize>) {
+        for item in &self.stack {
+            match item {
+                RouteTerm::LogicalOperator(_op) => {
+                    // need to do nothing here
+                }
+                RouteTerm::Predicate(p) => {
+                    let val = map.get_mut(&p.lhs.var_name).unwrap();
+                    *val -= 1;
+
+                    if *val == 0 {
+                        assert!(map.remove(&p.lhs.var_name).is_some());
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl Validate for Route {
```

Comment:
We should do this on AST instead of on Route.

Reply:
Route also need validation, in case there may be anything wrong with converion.